src/machine/io.c

```diff
@@ -248,30 +243,49 @@ static const char xdigits[16] = {
     "0123456789ABCDEF"
 };
 
-static char *fmt_x(word_t x, char *s, int lower)
+static char *fmt_x(uintmax_t x, char *s, int lower)
 {
     for (; x; x >>= 4) {
         *--s = xdigits[(x & 15)] | lower;
     }
     return s;
 }
 
-static char *fmt_o(word_t x, char *s)
+static char *fmt_o(uintmax_t x, char *s)
 {
     for (; x; x >>= 3) {
         *--s = '0' + (x & 7);
     }
     return s;
 }
 
-static char *fmt_u(word_t x, char *s)
+static char *fmt_u(uintmax_t x, char *s)
 {
-    unsigned long y;
-    for (; x > ULONG_MAX; x /= 10) {
-        *--s = '0' + x % 10;
-    }
-    for (y = x;           y; y /= 10) {
-        *--s = '0' + y % 10;
+    while (0 != x) {
+#if defined(CONFIG_ARCH_AARCH32) || defined(CONFIG_ARCH_RISCV32)
+        /* On 32-bit systems dividing a 64-bit number by 10 make the compiler
+         * call a helper functions. On ARM it's __aeabi_uldivmod(), on RISC-V
+         * it's __umoddi3(). These are not available in the kernel, thus the
+         * algorithm below is needed. Credits go to Hacker's Delight.
```

Comment:
How was the old version of this code working in the past then? It seems like the compiler was previously able to optimize the code into something that didn't require calling helper functions?

Reply:
Looks to me as if it has never run into code that has to divided a 64-bit value by a arbitrary non-power-of-2 value. Also, integers where limited to 32-bit on 32-bit platforms. I've found all this because I was trying to print a 64-bit time value and it turned out only the lower 32-bit were printed. Furthermore, the division issue just "really" hows up when printing a number in decimal.